packages/flutter/lib/src/rendering/object.dart

```diff
@@ -4751,726 +4501,1163 @@ mixin RelayoutWhenSystemFontsChangeMixin on RenderObject {
   }
 }
 
-/// Describes the semantics information a [RenderObject] wants to add to its
-/// parent.
-///
-/// It has two notable subclasses:
-///  * [_InterestingSemanticsFragment] describing actual semantic information to
-///    be added to the parent.
-///  * [_ContainerSemanticsFragment]: a container class to transport the semantic
-///    information of multiple [_InterestingSemanticsFragment] to a parent.
-abstract class _SemanticsFragment {
-  _SemanticsFragment({required this.dropsSemanticsOfPreviousSiblings});
-
-  /// Incorporate the fragments of children into this fragment.
-  void addAll(Iterable<_InterestingSemanticsFragment> fragments);
+/// Properties of _RenderObjectSemantics that are imposed from parent.
+@immutable
+final class _SemanticsParentData {
+  const _SemanticsParentData({
+    required this.mergeIntoParent,
+    required this.blocksUserActions,
+    required this.explicitChildNodes,
+    required this.tagsForChildren,
+  });
```

Comment:
For my own understanding: Why are the geometrics not part of the parent data? I thought those are passed down from the parent as well?

Reply:
geometric can only be determined when we know the shape of semantics tree.

the parentData is used for determine the shape of semantics tree, e.g. explicitChildNode, blocksUseractions.

basically we can't determine both geometry, parentData, and semantics tree shape in one pass.

we could merge them and make parentData mutable and do a two pass to update the parentData, but i felt that is harder to manage.

Reply:
+1 to keeping parentData immutable. Maybe leave this as a comment here that geometrics are not included because we use parentData to determine the shape of the tree and geometrics can only be calculated after the shape of the tree is clear.